lib/license.js

```diff
@@ -1,38 +1,95 @@
-var BSD_LICENSE = /edistribution and use in source and binary forms, with or withou/;
-var WTFPL_LICENSE = /DO WHAT THE FUCK YOU WANT TO PUBLIC LICENSE/;
-var ISC_LICENSE = /The ISC License/;
+var MIT_LICENSE = /ermission is hereby granted, free of charge, to any/i;
+var BSD_LICENSE = /edistribution and use in source and binary forms, with or withou/i;
+var WTFPL_LICENSE = /DO WHAT THE FUCK YOU WANT TO PUBLIC LICENSE/i;
+var ISC_LICENSE = /The ISC License/i;
+var APACHE_LICENSE = /Apache License/i;
 var MIT = /MIT\b/;
 var BSD = /BSD\b/;
 var ISC = /ISC\b/;
-var APACHE = /Apache License\b/;
+var APACHE = /Apache\b/;
 var WTFPL = /WTFPL\b/;
+var allNewlines = /\r?\n/g;
 
+var spdx = require('spdx-license-list/spdx-full');
+var cmp = require('leven');
+
+var licenses = Object.keys(spdx).map(function(key) {
+    var license = spdx[key];
+    if (!license.osiApproved) {
+        return false;
+    }
+
+    license.key = key;
+    return {
+        key: key,
+        name: license.name,
+        url: license.url,
+        license: license.license
+    };
+}).filter(Boolean);
+
+function bestMatch(str, filter) {
+    var tests = licenses, results;
+
+    if (filter) {
+        filter = filter.toLowerCase();
+        tests = licenses.filter(function(a) {
+            return a.name.toLowerCase().indexOf(filter) > -1;
+        });
+    }
+
+    results = tests.map(function(license) {
+        if (license.license.length * 2 < str) {
+            return [license.key, Infinity];
+        }
+        var dist = cmp(str, license.license);
+        return [license.key, dist];
+    });
+
+    results.sort(function(a, b) {
+        return a[1] - b[1];
+    });
+
+    return results;
+}
 
 module.exports = function(str) {
-    if (str) {
-        str = str.replace('\n', '');
+
+    if (!str || str.toLowerCase().indexOf('error:') > -1) {
+        return null;
     }
+
+    var filter = null, results;
+    str = str.replace(allNewlines, '');
+
+    // Attempt to setup a fast path to avoid extraneous compares
     if (typeof str === 'undefined' || !str) {
-        return 'Undefined';
+        return null;
     } else if (ISC_LICENSE.test(str)) {
-        return 'ISC*';
+        filter = 'ISC';
     } else if (MIT_LICENSE.test(str)) {
-        return 'MIT*';
+        filter = 'MIT';
     } else if (BSD_LICENSE.test(str)) {
-        return 'BSD*';
+        filter = 'BSD';
     } else if (WTFPL_LICENSE.test(str)) {
-        return 'WTFPL*';
+        filter = 'WTFPL';
+    } else if (APACHE_LICENSE.test(str)) {
+        filter = 'Apache';
     } else if (ISC.test(str)) {
-        return 'ISC*';
+        filter = 'ISC';
     } else if (MIT.test(str)) {
-        return 'MIT*';
+        filter = 'MIT';
     } else if (BSD.test(str)) {
-        return 'BSD*';
+        filter = 'BSD';
     } else if (WTFPL.test(str)) {
-        return 'WTFPL*';
+        filter = 'WTFPL';
     } else if (APACHE.test(str)) {
-        return 'Apache*';
+        filter = 'Apache';
+    } else {
```

Comment:
both of the cases you mention (and others) intentionally return `null`, which has not changed from the original.  The matches are then sent along to attempt a textual match using the guess as a spdx license filter (so it doesn't test the entire corpus of spdx license texts).